web/src/components/customer/listing-details/contexts/CommitContext.tsx

```diff
@@ -0,0 +1,140 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, {useContext, useState, useEffect} from 'react';
+
+import {getCommits, addCommit, deleteCommit} from 'api';
+import {useCustomerContext} from 'components/customer/contexts/CustomerContext';
+import {useCommitFeedbackPromptContext} from 'components/customer/listing-details/contexts/CommitFeedbackPromptContext';
+import {CommitStatus} from 'interfaces';
+
+type ContextType =
+  | {
+      commitStatus: CommitStatus | undefined;
+      onCommit: () => Promise<void>;
+      onUncommit: () => Promise<void>;
+    }
+  | undefined;
+
+const CommitContext = React.createContext<ContextType>(undefined);
+
+/**
+ * useContext hook that ensures it is used within a CommitContext.
+ */
+const useCommitContext = () => {
+  const context = useContext(CommitContext);
+  if (context === undefined) {
+    throw new Error('useCommitContext must be used within a CommitContext');
+  }
+  return context;
+};
+
+interface CommitContextProps {
+  listingId: number;
+}
+
+/**
+ * CommitContextProvider provider with stateful commit details.
+ */
+const CommitContextProvider: React.FC<CommitContextProps> = ({
+  children,
+  listingId,
+}) => {
+  const {
+    customer,
+    idToken,
+    getCustomerWithLogin,
+    refetchCustomer,
+  } = useCustomerContext();
+
+  const {onOpen: onOpenPrompt} = useCommitFeedbackPromptContext();
+
+  const [commitStatus, setCommitStatus] = useState<CommitStatus>();
+  const [commitId, setCommitId] = useState<number | undefined>();
+
+  useEffect(() => {
+    if (customer === undefined) {
+      setCommitStatus(undefined);
+      return;
+    }
+
+    const fetchCommit = async () => {
+      const [commit] = await getCommits({
+        listingId,
+        customerId: customer.id,
+      });
+      if (commit !== undefined) {
+        setCommitId(commit.id);
+        setCommitStatus(commit.commitStatus);
+      }
+    };
+    fetchCommit();
+  }, [listingId, customer]);
+
+  const onCommit = async () => {
+    const customer = await getCustomerWithLogin();
+
+    if (customer === undefined || idToken === undefined) {
+      // TODO: Handle case when user refuse to login even after prompted
+      return;
+    }
+
+    const commit = await addCommit(
+      {
+        listingId,
+        customerId: customer.id,
+      },
+      idToken
+    );
+    // TODO: Handle addition error
+    setCommitId(commit.id);
+    await refetchCustomer();
```

Comment:
do you need to refresh listing?

Reply:
@ifhuang 
Line 105 does `setCommitStatus(commit.commitStatus);` which updates commitStatus.
Listing refreshes itself on commitStatus change in this block of code (highlighted in orange):
![image](https://user-images.githubusercontent.com/25261058/88013888-24983b00-cb50-11ea-99c5-cdbdcd877217.png)